### ISSUE TITLE
Remove the 'was [always] called with [exactly]' assertion

### ISSUE THREAD

**Issue by papandreou** [unexpectedjs/unexpected-sinon#32]
See #29.

- [ ] deprecate `was called with` in the docs
- [ ] wait a while so users have had the chance to replace their uses of `was [always] called with [exactly]` with `to have a call satisfying`
- [ ] replace `was [always] called with [exactly]` with an assertion that always throws a `bubbleThrough`error telling how to switch to `to have a call satisfying`
- [ ] release a new major version

**Comment by SleepWalker** [comment]
Hello.

I have one question about `to have a call satisfying` and `was [always] called with [exactly]`. I can't find a way to assert "always" (e.g. all calls should satisfy some expectation) using `to have a call satisfying`.

How should I assert, that some spy was never called with specific arguments?

**Comment by papandreou** [comment]
@SleepWalker, would something like [this](https://github.com/unexpectedjs/unexpected-sinon/pull/40) work for asserting that all calls to a spy satisfy a spec?

> How should I assert, that some spy was never called with specific arguments

This is a separate question, right? It's already there as `<spy> was never called with <any*>` (undocumented), but since we're getting rid of that family of assertions I guess we'll want to add a `to have no calls satisfying` assertion. Is it something you need often?

**Comment by papandreou** [comment]
`to have no calls satisfying` could look something like this: https://github.com/unexpectedjs/unexpected-sinon/pull/41

**Comment by papandreou** [comment]
@SleepWalker Both of the above are in the just released unexpected-sinon 10.6.0:

* [to have all calls satisfying](http://unexpected.js.org/unexpected-sinon/assertions/spy/to-have-all-calls-satisfying/)
* [to have no calls satisfying](http://unexpected.js.org/unexpected-sinon/assertions/spy/to-have-no-calls-satisfying/)

I hope that helps :)

**Comment by SleepWalker** [comment]
Thank you. This is awesome!

> Is it something you need often?

Not so often, but some times you need to be sure, that your code won't do unneeded actions.

Actually, using the first assertion it may be a little bit trickier, but it should work like so:

```js
expect(spy, 'to have all calls satisfying', [
    expect.it('not to equal', ...)
]);
```

but `to have no calls satisfying` easier to understand

**Comment by papandreou** [comment]
Yeah, you shouldn't have to "invert" the criterion to achieve that :)

**Comment by alexjeffburke** [comment]
@papandreou we may have a potential roadblock - while looking over mocha test suite, there are uses of "when called with" that have been introduced.

**Comment by sunesimonsen** [comment]
@alexjeffburke did you mean: was called with?

**Comment by sunesimonsen** [comment]
What do we win by removing this assertion from the code? Maybe we could just remove the docs or state that it is deprecated.

**Comment by alexjeffburke** [comment]
@sunesimonsen yes I did mean that.

I guess we don't gain all that much - its just that particular assertion suffers some "surprise" factor as a result of the varargs signature. What I've seen at work is people unfortunately using it because it's in the plugin and happens to work. If we keep it in I would encourage a deprecation notice.

**Comment by papandreou** [comment]
It’s already marked as deprecated: 
http://unexpected.js.org/unexpected-sinon/assertions/spy/was-called-with/

**Comment by alexjeffburke** [comment]
@papandreou if we feel strongly about it I think that should be a runtime thing - but you're right re documentation that it's kind of working around people referring to examples in bigger projects without reading the docs. If we believe the docs are enough then maybe we should close this :)

**Comment by papandreou** [comment]
The documentation link was for Sune as a reminder of where we are in the process. I still think we should remove it as per the plan we made, but if everyone else changed their minds that's also fair ;)

**Comment by sunesimonsen** [comment]
@papandreou I'm okay with that, but it would a new major right?

**Comment by sunesimonsen** [comment]
I was more thinking in terms of removing the current documentation and only leave the deprecation warning.

**Comment by sunesimonsen** [comment]
We can start by doing that.

**Comment by papandreou** [comment]
@sunesimonsen, yes, at least :)

But if new `was called with`s are being introduced almost 3 years after we put that notice in the docs,, clearly our deprecation strategy hasn't been working very well. I think we should try something else to try to discourage its use before we proceed to remove it.

**Comment by papandreou** [comment]
Added the `console.warn` here: https://github.com/unexpectedjs/unexpected-sinon/issues/32

Sent PR to mocha here: https://github.com/mochajs/mocha/pull/3680/checks
